File: problem32/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        List<Integer> arr = new ArrayList<>();
        int size = 0;
        while (head != null) {
            arr.add(head.val);
            head = head.next;
            size++;
        }
        int maxsum = 0, i = 0;
        while (i < size/2) {
            maxsum = Math.max(maxsum, arr.get(i)+arr.get(size-i-1));
            i++;
        }
        return maxsum;
    }
}

File: problem32/Solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        prev, nex, fast, maxSum = None, head.next, head.next.next, 0
        # Reverse first-half of the linkedlist
        while fast:
            head.next = prev
            prev = head
            head = nex
            nex = nex.next
            fast = fast.next.next
        head.next = prev
        # traverse through 2 linkedlists
        while head:
            pairSum = head.val + nex.val
            maxSum = max(pairSum, maxSum)
            head = head.next
            nex = nex.next
        return maxSum
